CoViewingExample/AudioDevices/ExampleAVPlayerAudioDevice.m

```diff
@@ -0,0 +1,1132 @@
+        NSLog(@"Error activating AVAudioSession: %@", error);
+    }
+
+    // TODO: Set preferred input channels to 1?
+}
+
+- (AudioStreamBasicDescription)microphoneInputStreamDescription {
+    AudioStreamBasicDescription capturingFormatDescription = self.capturingFormat.streamDescription;
+    capturingFormatDescription.mBytesPerFrame = 2;
+    capturingFormatDescription.mBytesPerPacket = 2;
+    capturingFormatDescription.mChannelsPerFrame = 1;
+    return capturingFormatDescription;
+}
+
+- (AudioStreamBasicDescription)nonInterleavedStereoStreamDescription {
+    AudioStreamBasicDescription capturingFormatDescription = self.capturingFormat.streamDescription;
+    capturingFormatDescription.mBytesPerFrame = 2;
+    capturingFormatDescription.mBytesPerPacket = 2;
+    capturingFormatDescription.mChannelsPerFrame = 2;
+    capturingFormatDescription.mFormatFlags |= kAudioFormatFlagIsNonInterleaved;
+    return capturingFormatDescription;
+}
+
+- (OSStatus)setupAudioCapturer:(ExampleAVPlayerCapturerContext *)capturerContext {
+    UInt32 enableInput = capturerContext ? 1 : 0;
+    OSStatus status = AudioUnitSetProperty(_voiceProcessingIO, kAudioOutputUnitProperty_EnableIO,
+                                           kAudioUnitScope_Input, kInputBus, &enableInput,
+                                           sizeof(enableInput));
+
+    if (status != noErr) {
+        NSLog(@"Could not enable/disable input bus!");
+        AudioComponentInstanceDispose(_voiceProcessingIO);
+        _voiceProcessingIO = NULL;
+        return status;
+    } else if (!enableInput) {
+        // Input is not required.
+        return noErr;
+    }
+
+    // Request mono audio capture regardless of hardware.
+    AudioStreamBasicDescription capturingFormatDescription = [self microphoneInputStreamDescription];
+
+    // Our converter will interleave the mono microphone input and player audio in one stereo stream.
+    if (_captureConverter == NULL) {
+        AudioStreamBasicDescription sourceFormat = [self nonInterleavedStereoStreamDescription];
+        AudioStreamBasicDescription destinationFormat = [self.capturingFormat streamDescription];
+        OSStatus status = AudioConverterNew(&sourceFormat,
+                                            &destinationFormat,
+                                            &_captureConverter);
+        if (status != noErr) {
+            NSLog(@"Could not create capture converter! code: %d", status);
+            return status;
+        }
+    }
+
+    status = AudioUnitSetProperty(_voiceProcessingIO, kAudioUnitProperty_StreamFormat,
+                                  kAudioUnitScope_Output, kInputBus,
+                                  &capturingFormatDescription, sizeof(capturingFormatDescription));
+    if (status != noErr) {
+        NSLog(@"Could not set stream format on the input bus!");
+        AudioComponentInstanceDispose(_voiceProcessingIO);
+        _voiceProcessingIO = NULL;
```

Comment:
should we dispose mixer as well ? Probably we can have a strategy, setup mixer first, and if `_voiceProcessingIO` gets into an error, we can dispose mixer as well. WDYT?